docs/source/tutorials/knn_outlier.ipynb

```diff
@@ -0,0 +1,447 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "94a7bb8a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "txt_labels = {0: 'airplane', \n",
+    "              1: 'automobile', \n",
+    "              2: 'bird',\n",
+    "              3: 'cat', \n",
+    "              4: 'deer', \n",
+    "              5: 'dog', \n",
+    "              6: 'frog', \n",
+    "              7: 'horse', \n",
+    "              8:'ship', \n",
+    "              9:'truck'}\n",
+    "\n",
+    "def plot_images(X,y):\n",
+    "    plt.rcParams[\"figure.figsize\"] = (9,7)\n",
+    "    for i in range(15):\n",
+    "        # define subplot\n",
+    "        ax = plt.subplot(3,5,i+1)\n",
+    "        ax.set_title(txt_labels[int(y[i])])\n",
+    "        # plot raw pixel data\n",
+    "        ax.imshow(X[i])\n",
+    "    # show the figure\n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9dddf313",
+   "metadata": {},
+   "source": [
+    "Observe how there are only animals left in the training set `trainX` below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "486acf25",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_images(trainX,trainy)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7cf11eb8",
+   "metadata": {},
+   "source": [
+    "The test set on the other hand still visibily contains the non-animal images: `[ship, airplane, automobile, truck]`. If we consider `trainX` to be the representation of data, then these non-animal images in test dataset `testX` become outliers."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1d2fdd09",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_images(testX,testy)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "15f57bae",
+   "metadata": {},
+   "source": [
+    "## 3. Import a model and embeddings\n",
+    "The model we are importing comes from [timm](https://timm.fast.ai/), a deep-learning library collection of SOTA models and utilities. We pass in the images into the model to generate embeddings in the feature space used in outlier detection.\n",
+    "\n",
+    "The model itself is a `resnet50` but outlier detection can be done with any object capable of generating feature embeddings."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "afd151d5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "import timm\n",
+    "\n",
+    "# This cell can take ~1-2 mins\n",
+    "model = timm.create_model('resnet50', pretrained=True, num_classes=0) # download the model from timm\n",
+    "trainX_torch = torch.from_numpy(trainX.swapaxes(1,3)) # turn trainX into tensor and fix channel dimension\n",
+    "train_feature_embeddings = model(trainX_torch) # fit model to train images and get train feature embeddings\n",
```

Comment:
One reason for the RAM issues is that you're doing a single forward pass through the net with all of the data at once. This should be broken up into a loop over small mini-batches of the data and the results concatenated back into a single array (especially important for GPU which has less RAM)

Reply:
same for test data below, we always only want to forward pass a small mini-batch through the network, never an entire dataset